core/src/main/java/org/elasticsearch/common/geo/parsers/ShapeParser.java

```diff
@@ -0,0 +1,128 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.geo.parsers;
+
+import com.vividsolutions.jts.geom.Coordinate;
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.common.Explicit;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.geo.builders.ShapeBuilder;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.mapper.GeoShapeFieldMapper;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * static base class for implementing a shape parser
+ */
+public abstract class ShapeParser {
+    public static ParseField FIELD_TYPE = new ParseField("type");
+    public static ParseField FIELD_COORDINATES = new ParseField("coordinates");
+    public static ParseField FIELD_GEOMETRIES = new ParseField("geometries");
+    public static ParseField FIELD_ORIENTATION = new ParseField("orientation");
+
+    /**
+     * Create a new {@link ShapeBuilder} from {@link XContent}
+     * @param parser parser to read the GeoShape from
+     * @param shapeMapper document field mapper reference required for spatial parameters relevant
+     *                     to the shape construction process (e.g., orientation)
+     *                     todo: refactor to place build specific parameters in the SpatialContext
+     * @return {@link ShapeBuilder} read from the parser or null
+     *          if the parsers current token has been <code>null</code>
+     * @throws IOException if the input could not be read
+     */
+    public static ShapeBuilder parse(XContentParser parser, GeoShapeFieldMapper shapeMapper) throws IOException {
+        if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+            return null;
+        } if (parser.currentToken() == XContentParser.Token.START_OBJECT) {
+            return GeoJsonParser.parse(parser, shapeMapper);
+        }
+        throw new ElasticsearchParseException("shape must be an object consisting of type and coordinates");
+    }
+
+    /**
+     * Create a new {@link ShapeBuilder} from {@link XContent}
+     * @param parser parser to read the GeoShape from
+     * @return {@link ShapeBuilder} read from the parser or null
+     *          if the parsers current token has been <code>null</code>
+     * @throws IOException if the input could not be read
+     */
+    public static ShapeBuilder parse(XContentParser parser) throws IOException {
+        return parse(parser, null);
+    }
+
+    /**
+     * Node used to represent a tree of coordinates.
+     * <p>
+     * Can either be a leaf node consisting of a Coordinate, or a parent with
+     * children
+     */
+    public static class CoordinateNode implements ToXContentObject {
```

Comment:
Does it make sense to move this to its own class? If not, can access be limited so it can only be used in this package?

Reply:
I can split this into its own class under the `parsers` package. Its unique to parsing but also reference by `GeoShapeType`